### ISSUE TITLE
Curriculum concept for pre-sorting data for training

### ISSUE THREAD

**Issue by cameronjacobson** [tensorflow/tensorflow#99562]
### Issue type

Feature Request

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.20.0

### Custom code

No

### Current behavior?

I was reading about Mahalanobis Distance, it seems this would be a perfect use cases for testing this new concept.

### Standalone code to reproduce the issue

```shell
N/A
```

### Relevant log output

```shell

```

**Comment by j4b3-21** [comment]
Would it make sense to expose curriculum learning functionality (such as Mahalanobis-distance–based pre-sorting) as a public API in tf.data? For example, something like dataset.sort_by_difficulty(metric="mahalanobis"). Or do you see this as something that should remain a user-implemented preprocessing step outside TensorFlow?

**Comment by cameronjacobson** [comment]
@j4b3-21  Thanks for the comment.  My hunch is that it would be beneficial to integrate curriculum-based training into the public API, semantically differentiating it from a general "sort" function in the code and documentation.

This suggestion stems from two points:

1. From what I understand, curriculum-based training can significantly impact a model's performance.  So conceptually this kind of transformation will become valuable to most AI/ML engineers if it's not already.
2. Treating it as a distinct concept in the code, with the ability to inject custom callbacks, allows for future innovation and easier adoption of state-of-the-art algorithms as they’re discovered by researchers.

I acknowledge that this approach might not conceptually align with everyone's view of TensorFlow's design, but I also wonder where else this capability would be better placed if not within TensorFlow.

**Comment by j4b3-21** [comment]
I feel that at this point, it makes sense to add this to tensorflow/addons rather than core TensorFlow.
This way, the idea can be developed and iterated on without affecting the stability of the main API. Later, if it matures and proves broadly useful, it could be considered for transition into the core.

In the meantime, I’d be happy to help by drafting a prototype or RFC, so that the concept can be explored in more detail and refined with community feedback.

**Comment by cameronjacobson** [comment]
That makes sense.  I've only dug into the code of Tensorflow on a few occasions for specific reasons, so I'm not sure where to begin, but if you feel inspired by the idea I wouldn't have any objection to anyone contributing or taking ownership of this.  Given that it would be my first contribution to this project, I wouldn't really feel comfortable taking the lead, but will be happy to do what I can to assist (if needed).

I imagine a POC could probably be implemented pretty easily, if someone knows where it belongs in the code, since it looks like SciPy has an implementation of Mahalanobis Distance, and NumPy has what is needed for calculating the inverse of the covariance matrix. For the calculation, my understanding is that we wouldn't be taking the "distance" between 2 vectors within the training data.  Instead one of the vectors would need to be the "mean" vector representing the mean of the entire dataset.  That mean vector would be compared to each individual row of the dataset and then we'd sort by that distance value.

I do have a question about the AddOns project.  It looks like the website says the AddOns project has been deprecated?  Does that just mean it's not supported by the Tensorflow team anymore?  Or does it mean we'd need to change our target for this code?

https://www.tensorflow.org/addons

**Comment by cameronjacobson** [comment]
Perhaps if the "curriculum" concept is too high level maybe Keras would be a better fit for this idea?

**Comment by nileshau-afk** [comment]
Hi @cameronjacobson  ,

Thank you so much for filing this issue. We truly appreciate your support. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

we look forward to seeing your contributions!
